object/slicer/slicer.go

```diff
@@ -164,20 +163,44 @@ func slice(ctx context.Context, ow ObjectWriter, header object.Object, data io.R
 	objectPayloadLimit := childPayloadSizeLimit(opts)
 
 	var n int
-	bChunk := opts.payloadBuffer
-	if bChunk == nil {
-		bChunk = make([]byte, objectPayloadLimit)
-	}
 
 	writer, err := initPayloadStream(ctx, ow, header, signer, opts)
 	if err != nil {
 		return rootID, fmt.Errorf("init writter: %w", err)
 	}
 
+	var buf []byte
+	var buffered uint64
+
 	for {
-		n, err = data.Read(bChunk)
+		buffered = writer.rootMeta.length
+		if writer.withSplit {
+			buffered = writer.childMeta.length
+		}
+
+		if buffered == objectPayloadLimit && uint64(len(writer.payloadBuffer)) <= objectPayloadLimit {
+			// in this case, the read buffers are exhausted, and it is unclear whether there
+			// will be more data. We need to know this right away, because the format of the
+			// final objects depends on it. So read to temp buffer
+			buf = []byte{0}
+		} else {
+			payloadBufferLen := uint64(len(writer.payloadBuffer))
+			if payloadBufferLen > buffered {
+				buf = writer.payloadBuffer[buffered:]
+			} else {
+				if writer.extraPayloadBuffer == nil {
+					// TODO(#544): support external buffer pools
+					writer.extraPayloadBuffer = make([]byte, writer.payloadSizeLimit-payloadBufferLen)
+					buf = writer.extraPayloadBuffer
+				} else {
+					buf = writer.extraPayloadBuffer[buffered-payloadBufferLen:]
```

Comment:
if `extraPayloadBuffer` is not `nil`, we are sure it is longer than `buffered-payloadBufferLen`?

Reply:
yes cuz `buffered  <= writer.payloadSizeLimit` always